### DISCUSSION TITLE
Issue with the Sous Vide node

### DISCUSSION THREAD

**Discussion by Rodriweiss** [andrei-tatar/node-red-contrib-smartnora#296] [General]
Hello,

I have been trying to fix some issues with a kitchen thermometer I made to warn me when liquids reach a certain temperature. The code is actually doing what it's supposed to do, however, there are some issues with the device traits. For some reason, the Nora node will only accept temperature info as the thermostatTemperatureAmbient / thermostatTemperatureSetpoint traits, not the proper temperatureAmbientCelsius / temperatureSetpointCelsius traits. As per Google's documentation:

> "action.devices.traits.TemperatureControl - Trait for devices (other than thermostats) that support controlling temperature, either within or around the device.
> 
> This differs from the [TemperatureSetting](https://developers.home.google.com/cloud-to-cloud/traits/temperaturesetting) trait, which is specifically for thermostat-style controls. The TemperatureSetting trait represents ambient (room/outdoor) temperature and should not be used for controlling the temperature of a specific device. In order to control the temperature of a specific device, you must use the TemperatureControl trait."

I'm wondering if this is an issue on my code. Here is the export of the nodes:

[
    {
        "id": "61f86d740329192f",
        "type": "group",
        "z": "f7a1060f7191035d",
        "name": "Kitchen Thermometer",
        "style": {
            "label": true
        },
        "nodes": [
            "1bebd9319c2eed85",
            "bcf396d1f693a9f6",
            "f7bf43af4844402e",
            "aa50f07310d8e618",
            "04ace71a372ce030",
            "ff6045b559c19872",
            "9b3b96d6c72c9459",
            "05c088579744b4ca",
            "8d63544433c663fe",
            "74f24ba070416686"
        ],
        "x": 14,
        "y": 1199,
        "w": 1192,
        "h": 142
    },
    {
        "id": "1bebd9319c2eed85",
        "type": "ui_gauge",
        "z": "f7a1060f7191035d",
        "g": "61f86d740329192f",
        "name": "Kitchen Thermometer",
        "group": "5b8c9e45308bc8d4",
        "order": 0,
        "width": 0,
        "height": 0,
        "gtype": "gage",
        "title": "Kitchen Thermometer",
        "label": "Cº",
        "format": "{{value}}",
        "min": 0,
        "max": "100",
        "colors": [
            "#00b500",
            "#e6e600",
            "#ca3838"
        ],
        "seg1": "",
        "seg2": "",
        "diff": false,
        "className": "",
        "x": 160,
        "y": 1240,
        "wires": []
    },
    {
        "id": "bcf396d1f693a9f6",
        "type": "noraf-sousvide",
        "z": "f7a1060f7191035d",
        "g": "61f86d740329192f",
        "devicename": "Kitchen Helper",
        "roomhint": "Kitchen",
        "name": "Kitchen Helper",
        "onOffSupported": true,
        "onOffMode": "cq",
        "temperatureSupported": true,
        "temperatureMode": "cq",
        "temperatureUnit": "C",
        "temperatureStepSize": 1,
        "temperatureRangeMin": "0",
        "temperatureRangeMax": "120",
        "startStopSupported": true,
        "startStopPausable": true,
        "timerSupported": true,
        "timerMaxLimitSeconds": 3600,
        "timerMode": "cq",
        "topic": "kitchen/helper/pub",
        "passthru": true,
        "nora": "4da8f7a9995a0ceb",
        "x": 520,
        "y": 1300,
        "wires": [
            [
                "ff6045b559c19872",
                "9b3b96d6c72c9459"
            ]
        ]
    },
    {
        "id": "f7bf43af4844402e",
        "type": "mqtt in",
        "z": "f7a1060f7191035d",
        "g": "61f86d740329192f",
        "name": "Kitchen Thermometer",
        "topic": "kitchen/temperature/push",
        "qos": "2",
        "datatype": "json",
        "broker": "c45e2cc547ba88b2",
        "nl": false,
        "rap": true,
        "rh": 0,
        "inputs": 0,
        "x": 140,
        "y": 1300,
        "wires": [
            [
                "1bebd9319c2eed85",
                "aa50f07310d8e618"
            ]
        ]
    },
    {
        "id": "aa50f07310d8e618",
        "type": "function",
        "z": "f7a1060f7191035d",
        "g": "61f86d740329192f",
        "name": "",
        "func": "tempTherm = msg.payload;\n\n\nmsg.payload = {\n    thermostatTemperatureAmbient: parseInt(tempTherm),\n        }\nreturn msg",
        "outputs": 1,
        "timeout": "",
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [],
        "x": 340,
        "y": 1300,
        "wires": [
            [
                "bcf396d1f693a9f6"
            ]
        ]
    },
    {
        "id": "04ace71a372ce030",
        "type": "function",
        "z": "f7a1060f7191035d",
        "g": "61f86d740329192f",
        "name": "Temp Check (old)",
        "func": "var temp1 = flow.get('temp1');\nvar temp2 = flow.get('temp2');\n\n\nwhile (temp1 == temp2) {\n    msg.payload = \"The desired temperature has been reached\";\n    return msg;\n    }\nreturn;\n\n\n\n\n\n\n\n\n   ",
        "outputs": 1,
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [],
        "x": 950,
        "y": 1300,
        "wires": [
            [
                "8d63544433c663fe"
            ]
        ]
    },
    {
        "id": "ff6045b559c19872",
        "type": "function",
        "z": "f7a1060f7191035d",
        "g": "61f86d740329192f",
        "name": "Store temp2",
        "func": "var temp2 = flow.get('temp2');\nvar isTemp2NaN = isNaN(msg.payload.thermostatTemperatureSetpoint);\n\nif (!isTemp2NaN){\n\ntemp2 = parseInt(msg.payload.thermostatTemperatureSetpoint);\n\nflow.set('temp2', temp2);\n\nmsg.payload = temp2;\n\nreturn msg;\n\n\n}\nreturn;",
        "outputs": 1,
        "timeout": "",
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [],
        "x": 730,
        "y": 1300,
        "wires": [
            [
                "04ace71a372ce030"
            ]
        ]
    },
    {
        "id": "9b3b96d6c72c9459",
        "type": "function",
        "z": "f7a1060f7191035d",
        "g": "61f86d740329192f",
        "name": "Store Temp1",
        "func": "var temp1 = flow.get('temp1');\n\ntemp1 = parseInt(msg.payload.thermostatTemperatureAmbient);\n\nflow.set('temp1', temp1);\n\nmsg.payload = temp1;\n\nreturn msg;\n",
        "outputs": 1,
        "timeout": "",
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [],
        "x": 730,
        "y": 1240,
        "wires": [
            [
                "05c088579744b4ca"
            ]
        ]
    },
    {
        "id": "05c088579744b4ca",
        "type": "rbe",
        "z": "f7a1060f7191035d",
        "g": "61f86d740329192f",
        "name": "",
        "func": "rbe",
        "gap": "",
        "start": "",
        "inout": "out",
        "septopics": true,
        "property": "payload",
        "topi": "topic",
        "x": 870,
        "y": 1240,
        "wires": [
            [
                "04ace71a372ce030"
            ]
        ]
    },
    {
        "id": "8d63544433c663fe",
        "type": "cast-to-client",
        "z": "f7a1060f7191035d",
        "g": "61f86d740329192f",
        "name": "Cast",
        "url": "",
        "contentType": "",
        "message": "",
        "language": "en",
        "ip": "192.168.100.135",
        "port": "8009",
        "volume": "",
        "x": 1130,
        "y": 1300,
        "wires": [
            []
        ]
    },
    {
        "id": "74f24ba070416686",
        "type": "mqtt in",
        "z": "f7a1060f7191035d",
        "g": "61f86d740329192f",
        "name": "On or Off",
        "topic": "kitchen/temperature/push",
        "qos": "2",
        "datatype": "utf8",
        "broker": "c45e2cc547ba88b2",
        "nl": false,
        "rap": true,
        "rh": 0,
        "inputs": 0,
        "x": 340,
        "y": 1240,
        "wires": [
            [
                "bcf396d1f693a9f6"
            ]
        ]
    },
    {
        "id": "5b8c9e45308bc8d4",
        "type": "ui_group",
        "name": "Sensors",
        "tab": "bd1610225d88de3a",
        "order": 2,
        "disp": true,
        "width": "6",
        "collapse": false,
        "className": ""
    },
    {
        "id": "4da8f7a9995a0ceb",
        "type": "noraf-config",
        "name": "Nora Devices",
        "group": "Devices",
        "twofactor": "off",
        "twofactorpin": "",
        "localexecution": true,
        "structure": "",
        "storeStateInContext": false,
        "disableValidationErrors": false,
        "sendDeviceNameAndLocation": false
    },
    {
        "id": "c45e2cc547ba88b2",
        "type": "mqtt-broker",
        "name": "localhost:1883",
        "broker": "localhost",
        "port": "1883",
        "clientid": "",
        "usetls": false,
        "protocolVersion": "5",
        "keepalive": "60",
        "cleansession": true,
        "birthTopic": "",
        "birthQos": "2",
        "birthRetain": "true",
        "birthPayload": "",
        "birthMsg": {},
        "closeTopic": "",
        "closeQos": "0",
        "closePayload": "",
        "closeMsg": {},
        "willTopic": "",
        "willQos": "0",
        "willPayload": "",
        "willMsg": {},
        "sessionExpiry": ""
    },
    {
        "id": "bd1610225d88de3a",
        "type": "ui_tab",
        "name": "My Home",
        "icon": "dashboard",
        "order": 1,
        "disabled": false,
        "hidden": false
    }
]

**Comment by andrei-tatar** [comment]
@Rodriweiss Seems the sousvide node is implemented using the `TemperatureSetting` trait:

https://github.com/andrei-tatar/node-red-contrib-smartnora/blob/edb1bf9c347ff69a183dcba86b7e4980327c253e/src/nodes/nora-sousvide.ts#L52

I honestly don't remember why it was done like this. I will have a look at updating it.

LE: Checked the code history and it seems it was initially added as `TemperatureControl`. Asked Toby if he can remember the reason:
https://github.com/andrei-tatar/node-red-contrib-smartnora/pull/133

**Reply by Toby-Mills** [reply]
@andrei-tatar @Rodriweiss 
I don't remember for sure why I used that trait, as TemperatureControl does sound like the correct one to use. I think it was possibly related to the different UI's that Google Home presented for controlling the different traits.

**Reply by Rodriweiss** [reply]
Afaik Google Home doesn't have UI controls for Sous Vide devices yet. The issue I'm having is very minor, as I said, the device and nodes work, the issue is that because it is half registering as a thermostat in Google Home, when I ask Google about the temperature in the house it will first get the status from my proper thermostat and then try to get info from the sous vide device

**Reply by Toby-Mills** [reply]
The UI Controls are based on the traits, so I can set the target temperature on my sous vide using the same UI that Google Home uses for a thermostat.
If you want to change the code to rather use the TemperatureControl trait instead, that's fine by me as it does seem to be the correct one to use. Anybody using that device type in Node Red would need to change their flows to accommodate the change. @andrei-tatar what do you think?

**Reply by Rodriweiss** [reply]
It seems the node must be kept as is for the touch controls to work. TemperatureControl does not have buttons enabled in GH yet afterall.

https://developers.home.google.com/cloud-to-cloud/features/touch-controls

I'm fine with keeping it as is until proper controls are implemented, it's just a minor issue

**Reply by andrei-tatar** [reply]
@Toby-Mills I was thinking there are 2 options:
1. switch to TemperatureControl but keep the in/out properties backwards compatible. This will break the UI control from Home app
2. add a dropdown option to choose which trait to use. They seem very similar anyways.

Probably best option is 2. I can look into next week.

**Reply by Rodriweiss** [reply]
@andrei-tatar changing the subject, would it be possible to add a node for media remote control?https://developers.home.google.com/cloud-to-cloud/guides/remotecontrol 

Seems like this could be useful as way to control IR devices that don't fit the existing categories

**Reply by andrei-tatar** [reply]
@Rodriweiss  it's already there, under the `media` node. Just choose the appropriate `Device Type` from the dropdown.
